File: backend/Services/Validation/MediaValidationRules.cs
using PostPilot.Api.Enums;

namespace PostPilot.Api.Services.Validation;

/// <summary>
/// Platform and placement-specific media validation rules.
/// This is the single source of truth for all media validation constraints.
/// </summary>
public static class MediaValidationRules
{
    /// <summary>
    /// Gets the validation rules for a specific platform, placement, and media type combination.
    /// Returns null if no rules are defined for the combination.
    /// </summary>
    /// <param name="carousel">
    /// When true, the media item is part of a multi-item carousel and any carousel-specific
    /// override is used (currently only Instagram Feed video, whose per-item duration cap is 60s
    /// instead of the 180s single-video cap). When no override exists for the combination, the
    /// normal single-item rule is returned — image rules are identical whether single or carousel.
    /// </param>
    public static MediaValidationRule? GetRules(Platform platform, Placement placement, MediaType mediaType, bool carousel = false)
    {
        var key = (platform, placement, mediaType);

        if (carousel && CarouselItemOverrides.TryGetValue(key, out var carouselRule))
            return carouselRule;

        return Rules.TryGetValue(key, out var rule) ? rule : null;
    }

    /// <summary>
    /// Checks if rules exist for a specific combination.
    /// </summary>
    public static bool HasRules(Platform platform, Placement placement, MediaType mediaType)
    {
        return Rules.ContainsKey((platform, placement, mediaType));
    }

    /// <summary>
    /// Carousel-item overrides keyed by (Platform, Placement, MediaType). Used ONLY when a media
    /// item is part of a multi-item carousel and the carousel rule differs from the single-item
    /// rule. Combinations absent here fall back to <see cref="Rules"/> (i.e. carousel images use
    /// the same rule as single images).
    ///
    /// <para>Currently the only difference is Instagram Feed VIDEO duration: a single Feed video
    /// may run up to 180s (published by Meta as a Reel), but a video INSIDE a Feed carousel is
    /// capped at 60s. Everything else (MP4/MOV, 50MB, 3s minimum, readability, no codec/fps/
    /// dimension/aspect checks) is identical to the single-video rule.</para>
    /// </summary>
    private static readonly Dictionary<(Platform, Placement, MediaType), MediaValidationRule> CarouselItemOverrides = new()
    {
        [(Platform.Instagram, Placement.Feed, MediaType.Video)] = new MediaValidationRule
        {
            AllowedMimeTypes = ["video/mp4", "video/quicktime"],
            AllowedContainers = ["mp4", "mov"],
            MaxBytes = 50L * 1024 * 1024, // 50MB — same product cap as a single Feed video
            DurationMinSeconds = 3,
            DurationMaxSeconds = 60, // carousel video items: 3–60s (single Feed video allows up to 180s)
        },
    };

    /// <summary>
    /// All validation rules keyed by (Platform, Placement, MediaType).
    /// </summary>
    private static readonly Dictionary<(Platform, Placement, MediaType), MediaValidationRule> Rules = new()
    {
        // ============================================
        // FACEBOOK PAGE - FEED
        // ============================================
        // Facebook Feed Image Rules
        // Source: https://developers.facebook.com/docs/graph-api/reference/page/photos/
        [(Platform.Facebook, Placement.Feed, MediaType.Image)] = new MediaValidationRule
        {
            // Final product policy: JPG/JPEG + PNG only (no GIF/BMP/TIFF/WebP — they are
            // neither advertised in the UI nor accepted by upload init).
            AllowedMimeTypes = ["image/jpeg", "image/png"],
            MaxBytes = 10L * 1024 * 1024, // 10MB — MVP-supported limit, aligned with the current Facebook Page photo cap
            MinWidth = 320,
            MinHeight = 320,
            MaxWidth = 2048, // Recommended max (larger images are resized)
            MaxHeight = 2048,
            AspectRatioMin = 0.5625, // 9:16 (portrait)
            AspectRatioMax = 1.91, // ~1.91:1 (landscape)
            QualityWarningMinWidth = 600,
            QualityWarningMinHeight = 600,
        },

        // Facebook Feed Video Rules
        // Source: https://developers.facebook.com/docs/video-api/getting-started
        [(Platform.Facebook, Placement.Feed, MediaType.Video)] = new MediaValidationRule
        {
            // Final product policy: MP4 + MOV only. MOV (video/quicktime) stays supported for
            // iPhone compatibility, but its codec/audio must still pass when metadata is
            // available. AVI/WebM are intentionally NOT supported (no upload/publisher support).
            AllowedMimeTypes = ["video/mp4", "video/quicktime"],
            AllowedContainers = ["mp4", "mov"],
            AllowedVideoCodecs = ["h264", "hevc"],
            AllowedAudioCodecs = ["aac"],
            MaxBytes = 50L * 1024 * 1024, // 50MB — product cap for Facebook Feed video, matching the Supabase Free global upload limit
            MinWidth = 120,
            MinHeight = 120,
            MaxWidth = 4096,
            MaxHeight = 4096,
            AspectRatioMin = 0.5625, // 9:16 (portrait)
            AspectRatioMax = 1.91, // ~1.91:1 (landscape)
            DurationMinSeconds = 3, // product/MVP: small social videos only
            DurationMaxSeconds = 180, // product/MVP duration cap, NOT Meta's maximum (Meta allows hours)
            MaxFps = 60,
            RecommendedWidth = 1280,
            RecommendedHeight = 720,
        },

        // ============================================
        // FACEBOOK PAGE - STORY
        // ============================================
        // Facebook Story Image Rules
        // Facebook Story images are validated ONLY for supported type + file size. Facebook
        // Stories have NO dimension, resolution, orientation, or aspect-ratio requirements in
        // this product: a square, landscape, tiny, huge, wide, or tall image is accepted as long
        // as it is a valid JPG/PNG within the size cap. Deliberately NO Min/Max width/height, no
        // aspect-ratio range, no preferred 9:16, no quality warning, no recommended dimensions.
        [(Platform.Facebook, Placement.Story, MediaType.Image)] = new MediaValidationRule
        {
            // Final product policy: JPG/JPEG + PNG only.
            AllowedMimeTypes = ["image/jpeg", "image/png"],
            MaxBytes = 10L * 1024 * 1024, // 10MB — unchanged Facebook Story image cap
        },

        // Facebook Story Video Rules
        // Facebook Story videos are validated ONLY for container/type, file size, duration, and
        // readability. NO dimension, resolution, orientation, aspect-ratio, frame-rate, OR CODEC
        // requirements: a readable MP4/MOV within the size + duration limits is accepted whatever
        // its video/audio codec (including no audio stream, and unknown/missing codec names), and
        // Meta decides at publish time whether an uncommon codec combination is playable.
        // Codec validation is deliberately omitted (no AllowedVideoCodecs / AllowedAudioCodecs →
        // the engine skips both codec checks). FPS was likewise removed. Container + size +
        // duration + metadata readability remain fully enforced below.
        [(Platform.Facebook, Placement.Story, MediaType.Video)] = new MediaValidationRule
        {
            // Final product policy: MP4 + MOV container only (MOV for iPhone compatibility).
            AllowedMimeTypes = ["video/mp4", "video/quicktime"],
            AllowedContainers = ["mp4", "mov"],
            MaxBytes = 50L * 1024 * 1024, // 50MB (exactly 52,428,800 bytes) — Facebook Story video product cap
            DurationMinSeconds = 3, // supported duration range: Facebook Story videos are 3-90 seconds (inclusive)
            DurationMaxSeconds = 90,
        },

        // ============================================
        // INSTAGRAM - FEED
        // ============================================
        // Instagram Feed Image Rules — FINALIZED product policy (see spec).
        // Hard rules ONLY: accepted format, 8MB cap, and aspect ratio 4:5–1.91:1 (inclusive).
        // Deliberately NO dimension rules (min/max width/height), NO recommended resolution, NO
        // low-resolution/quality warnings, and NO cross-item aspect matching. Meta downscales
        // large images itself, so there is no platform width/height limit to enforce here.
        //
        // Meta accepts JPEG ONLY for Instagram, so this native rule stays JPEG-only; a PNG upload
        // is validated against its Instagram JPEG derivative instead (see EffectiveMediaResolver),
        // so .jpg/.jpeg/.png uploads are all supported end-to-end. (Any general decode/
        // decompression-bomb protection lives in the image decoder, not here, and is not an
        // Instagram platform dimension rule.)
        [(Platform.Instagram, Placement.Feed, MediaType.Image)] = new MediaValidationRule
        {
            AllowedMimeTypes = ["image/jpeg"],
            MaxBytes = 8L * 1024 * 1024, // 8MB — Instagram platform limit
            AspectRatioMin = 0.8, // 4:5 (portrait), inclusive
            AspectRatioMax = 1.91, // 1.91:1 (landscape), inclusive
        },

        // Instagram Feed Video Rules (single video) — FINALIZED product policy (see spec).
        // A single IG feed video is published by Meta as a Reel, but stays user-facing "Feed".
        // Hard rules ONLY: MP4/MOV container, 50MB cap, 3–180s duration, readable video stream.
        // Deliberately NO codec (H.264/HEVC), NO audio codec, NO frame-rate, NO dimensions, and
        // NO aspect-ratio prevalidation — Meta decides at publish time whether an unusual encoding
        // is playable. (The shared inspector may still READ codec/fps/dimensions for metadata; a
        // missing AllowedVideoCodecs/AllowedAudioCodecs list means the engine skips those checks,
        // and null dimensions/aspect skip the dimension and aspect checks.)
        [(Platform.Instagram, Placement.Feed, MediaType.Video)] = new MediaValidationRule
        {
            AllowedMimeTypes = ["video/mp4", "video/quicktime"], // MOV for iPhone compatibility
            AllowedContainers = ["mp4", "mov"],
            MaxBytes = 50L * 1024 * 1024, // 50MB — product cap for Instagram Feed video
            DurationMinSeconds = 3,
            DurationMaxSeconds = 180, // single Feed video; carousel video items are capped at 60s (see CarouselItemOverrides)
        },

        // ============================================
        // INSTAGRAM - STORY
        // ============================================
        // Instagram Story Image Rules
        // Source: https://developers.facebook.com/docs/instagram-platform/instagram-graph-api/reference/ig-user/media
        [(Platform.Instagram, Placement.Story, MediaType.Image)] = new MediaValidationRule
        {
            // Meta accepts JPEG ONLY for Instagram. This rule is the NATIVE publish format;
            // a PNG upload is validated against its Instagram JPEG derivative instead (see
            // EffectiveMediaResolver), so PNG users are not rejected for being non-JPEG.
            AllowedMimeTypes = ["image/jpeg"],
            MaxBytes = 8L * 1024 * 1024, // 8MB — Instagram platform limit
            MinWidth = 320,
            MinHeight = 320,
            MaxWidth = 1080,
            MaxHeight = 1920,
            AspectRatioMin = 0.50,
            AspectRatioMax = 0.75,
            PreferredAspectRatio = 0.5625, // 9:16
            AspectRatioWarningTolerance = 0.02,
            QualityWarningMinWidth = 600,
            QualityWarningMinHeight = 600,
            RecommendedWidth = 1080,
            RecommendedHeight = 1920,
        },

        // Instagram Story Video Rules
        [(Platform.Instagram, Placement.Story, MediaType.Video)] = new MediaValidationRule
        {
            // MP4 + MOV (MOV for iPhone compatibility). H.264 or HEVC/H.265, AAC audio.
            AllowedMimeTypes = ["video/mp4", "video/quicktime"],
            AllowedContainers = ["mp4", "mov"],
            AllowedVideoCodecs = ["h264", "hevc"],
            AllowedAudioCodecs = ["aac"],
            MaxBytes = 50L * 1024 * 1024, // 50MB — product cap for Instagram Story video
            MinWidth = 320,
            MinHeight = 320,
            MaxWidth = 1080,
            MaxHeight = 1920,
            AspectRatioMin = 0.50,
            AspectRatioMax = 0.75,
            PreferredAspectRatio = 0.5625, // 9:16
            AspectRatioWarningTolerance = 0.02,
            DurationMinSeconds = 3, // Meta/platform limit: story videos are 3-60 seconds
            DurationMaxSeconds = 60,
            MinFps = 23,
            MaxFps = 60,
            RecommendedWidth = 1080,
            RecommendedHeight = 1920,
        },

        // ============================================
        // TWITTER/X - FEED (for future implementation)
        // ============================================
        [(Platform.Twitter, Placement.Feed, MediaType.Image)] = new MediaValidationRule
        {
            AllowedMimeTypes = ["image/jpeg", "image/png", "image/gif", "image/webp"],
            MaxBytes = 5L * 1024 * 1024, // 5MB (15MB for GIFs)
            MinWidth = 100,
            MinHeight = 100,
            MaxWidth = 4096,
            MaxHeight = 4096,
            AspectRatioMin = 0.5, // 1:2
            AspectRatioMax = 3.0, // 3:1
            RecommendedWidth = 1200,
            RecommendedHeight = 675,
        },

        [(Platform.Twitter, Placement.Feed, MediaType.Video)] = new MediaValidationRule
        {
            AllowedMimeTypes = ["video/mp4"],
            AllowedContainers = ["mp4"],
            AllowedVideoCodecs = ["h264"],
            AllowedAudioCodecs = ["aac"],
            MaxBytes = 512L * 1024 * 1024, // 512MB
            MinWidth = 32,
            MinHeight = 32,
            MaxWidth = 1920,
            MaxHeight = 1200,
            AspectRatioMin = 0.5, // 1:2
            AspectRatioMax = 2.0, // 2:1
            DurationMinSeconds = 0.5,
            DurationMaxSeconds = 140, // 2 minutes 20 seconds
            MinFps = 25,
            MaxFps = 60,
            RecommendedWidth = 1280,
            RecommendedHeight = 720,
        },

        // ============================================
        // LINKEDIN - FEED (for future implementation)
        // ============================================
        [(Platform.LinkedIn, Placement.Feed, MediaType.Image)] = new MediaValidationRule
        {
            AllowedMimeTypes = ["image/jpeg", "image/png", "image/gif"],
            MaxBytes = 8L * 1024 * 1024, // 8MB
            MinWidth = 276,
            MinHeight = 276,
            MaxWidth = 4320, // 36MP max
            MaxHeight = 4320,
            AspectRatioMin = 0.57, // ~9:16
            AspectRatioMax = 3.0, // 3:1
            RecommendedWidth = 1200,
            RecommendedHeight = 627,
        },

        [(Platform.LinkedIn, Placement.Feed, MediaType.Video)] = new MediaValidationRule
        {
            AllowedMimeTypes = ["video/mp4", "video/quicktime", "video/x-msvideo"],
            AllowedContainers = ["mp4", "mov", "avi"],
            AllowedVideoCodecs = ["h264"],
            AllowedAudioCodecs = ["aac", "mp3"],
            MaxBytes = 200L * 1024 * 1024, // 200MB for standard, 5GB for Premium
            MinWidth = 256,
            MinHeight = 144,
            MaxWidth = 4096,
            MaxHeight = 2304,
            AspectRatioMin = 0.5625, // 9:16
            AspectRatioMax = 2.4, // 2.4:1
            DurationMinSeconds = 3,
            DurationMaxSeconds = 600, // 10 minutes (30 minutes for some accounts)
            MinFps = 10,
            MaxFps = 60,
            RecommendedWidth = 1920,
            RecommendedHeight = 1080,
        },
    };
}

/// <summary>
/// Represents validation rules for a specific platform/placement/media type combination.
/// </summary>
public class MediaValidationRule
{
    // File type constraints
    public required string[] AllowedMimeTypes { get; init; }
    public string[]? AllowedContainers { get; init; } // For videos: mp4, mov, etc.
    public string[]? AllowedVideoCodecs { get; init; } // h264, hevc, vp9, etc.
    public string[]? AllowedAudioCodecs { get; init; } // aac, mp3, etc.

    // Size constraints
    public long MaxBytes { get; init; }

    // Dimension constraints. Null means "no dimension rule for this combination": the engine
    // skips the corresponding min/max check entirely (e.g. Facebook Story media, which has no
    // dimension, resolution, or orientation requirements). Non-null values are enforced as hard
    // limits (unless MaxWidthIsAdvisory downgrades the max to a warning).
    public int? MinWidth { get; init; }
    public int? MinHeight { get; init; }
    public int? MaxWidth { get; init; }
    public int? MaxHeight { get; init; }

    /// <summary>
    /// When true, exceeding <see cref="MaxWidth"/> produces a WARNING instead of a
    /// hard error, because the platform downscales oversized images rather than
    /// rejecting them (Instagram scales any width &gt; 1440px down to 1440px).
    /// Default false: most platforms reject oversized media outright.
    /// </summary>
    public bool MaxWidthIsAdvisory { get; init; }

    // Aspect ratio constraints (width / height). Null Min/Max means "no aspect-ratio rule": the
    // engine skips the range check AND any preferred-ratio warning (e.g. Facebook Story media,
    // which has no aspect-ratio requirement).
    public double? AspectRatioMin { get; init; }
    public double? AspectRatioMax { get; init; }
    public double? PreferredAspectRatio { get; init; }
    public double? AspectRatioWarningTolerance { get; init; }

    // Video-specific constraints
    public double? DurationMinSeconds { get; init; }
    public double? DurationMaxSeconds { get; init; }
    public int? MinFps { get; init; }
    public int? MaxFps { get; init; }

    // Recommendations (for warnings, not errors)
    public int? RecommendedWidth { get; init; }
    public int? RecommendedHeight { get; init; }
    public int? QualityWarningMinWidth { get; init; }
    public int? QualityWarningMinHeight { get; init; }
}

File: frontend/src/constants/mediaValidationRules.ts
/**
 * Media validation rules for client-side pre-checks.
 * NOTE: Backend is the authoritative source of truth for validation rules.
 * These are kept minimal for UX pre-checks before upload.
 * If rules change, update backend first, then sync here.
 */

import type { PlatformId } from './validationLimits'

export type MediaType = 'Image' | 'Video'
export type Placement = 'Feed' | 'Story' | 'Reel'

/** Validation status returned from backend */
export type ValidationStatus = 'Pending' | 'Valid' | 'Invalid' | 'Warning'

/** Validation error from backend */
export interface MediaValidationError {
  code: string
  field: string
  message: string
  expected: string | null
  actual: string | null
}

/** Validation warning from backend */
export interface MediaValidationWarning {
  code: string
  field: string
  message: string
  recommendation: string | null
}

/** Extracted metadata from backend */
export interface ExtractedMediaMetadata {
  width: number | null
  height: number | null
  durationSeconds: number | null
  aspectRatio: number | null
  mimeType: string | null
  sizeBytes: number | null
  container: string | null
  videoCodec: string | null
  audioCodec: string | null
  fps: number | null
}

/** Validation result from backend */
export interface MediaValidationResult {
  status: ValidationStatus
  errors: MediaValidationError[]
  warnings: MediaValidationWarning[]
  metadata: ExtractedMediaMetadata | null
}

/** Client-side validation rule (subset of backend rules) */
export interface ClientMediaValidationRule {
  allowedMimeTypes: string[]
  maxBytes: number
  minWidth: number
  minHeight: number
  maxWidth: number
  maxHeight: number
  aspectRatioMin: number
  aspectRatioMax: number
  durationMinSeconds?: number
  durationMaxSeconds?: number
}

/** Platform-specific rules key */
type RuleKey = `${PlatformId}:${Lowercase<Placement>}:${Lowercase<MediaType>}`

/**
 * Client-side validation rules (matches backend MediaValidationRules.cs).
 * Keep this in sync with backend - backend is authoritative.
 */
const clientValidationRules: Partial<Record<RuleKey, ClientMediaValidationRule>> = {
  // Facebook Feed Image
  'facebook:feed:image': {
    allowedMimeTypes: ['image/jpeg', 'image/png', 'image/gif', 'image/bmp', 'image/tiff', 'image/webp'],
    maxBytes: 4 * 1024 * 1024, // 4MB
    minWidth: 320,
    minHeight: 320,
    maxWidth: 2048,
    maxHeight: 2048,
    aspectRatioMin: 0.5625, // 9:16
    aspectRatioMax: 1.91, // ~1.91:1
  },

  // Facebook Feed Video
  'facebook:feed:video': {
    allowedMimeTypes: ['video/mp4', 'video/quicktime', 'video/x-msvideo', 'video/webm'],
    maxBytes: 1024 * 1024 * 1024, // 1GB
    minWidth: 120,
    minHeight: 120,
    maxWidth: 4096,
    maxHeight: 4096,
    aspectRatioMin: 0.5625,
    aspectRatioMax: 1.91,
    durationMinSeconds: 1,
    durationMaxSeconds: 240 * 60, // 4 hours
  },

  // Facebook Story Image
  'facebook:story:image': {
    allowedMimeTypes: ['image/jpeg', 'image/png', 'image/gif', 'image/bmp', 'image/tiff', 'image/webp'],
    maxBytes: 4 * 1024 * 1024, // 4MB
    minWidth: 320,
    minHeight: 320,
    maxWidth: 1080,
    maxHeight: 1920,
    aspectRatioMin: 0.5625, // 9:16
    aspectRatioMax: 0.5625,
  },

  // Facebook Story Video
  'facebook:story:video': {
    allowedMimeTypes: ['video/mp4', 'video/quicktime'],
    maxBytes: 1024 * 1024 * 1024, // 1GB
    minWidth: 320,
    minHeight: 320,
    maxWidth: 1080,
    maxHeight: 1920,
    aspectRatioMin: 0.5625,
    aspectRatioMax: 0.5625,
    durationMinSeconds: 1,
    durationMaxSeconds: 120, // 2 minutes
  },

  // Instagram Feed Image
  'instagram:feed:image': {
    allowedMimeTypes: ['image/jpeg', 'image/png'],
    maxBytes: 8 * 1024 * 1024, // 8MB
    minWidth: 320,
    minHeight: 320,
    maxWidth: 1440,
    maxHeight: 1440,
    aspectRatioMin: 0.8, // 4:5
    aspectRatioMax: 1.91,
  },

  // Instagram Feed Video
  'instagram:feed:video': {
    allowedMimeTypes: ['video/mp4', 'video/quicktime'],
    maxBytes: 100 * 1024 * 1024, // 100MB
    minWidth: 500,
    minHeight: 500,
    maxWidth: 1920,
    maxHeight: 1920,
    aspectRatioMin: 0.8,
    aspectRatioMax: 1.91,
    durationMinSeconds: 3,
    durationMaxSeconds: 60, // 60 seconds
  },

  // Instagram Story Image
  'instagram:story:image': {
    allowedMimeTypes: ['image/jpeg', 'image/png'],
    maxBytes: 8 * 1024 * 1024, // 8MB
    minWidth: 320,
    minHeight: 320,
    maxWidth: 1080,
    maxHeight: 1920,
    aspectRatioMin: 0.5625, // 9:16
    aspectRatioMax: 0.5625,
  },

  // Instagram Story Video
  'instagram:story:video': {
    allowedMimeTypes: ['video/mp4', 'video/quicktime'],
    maxBytes: 100 * 1024 * 1024, // 100MB
    minWidth: 320,
    minHeight: 320,
    maxWidth: 1080,
    maxHeight: 1920,
    aspectRatioMin: 0.5625,
    aspectRatioMax: 0.5625,
    durationMinSeconds: 3,
    durationMaxSeconds: 60, // 60 seconds
  },

  // Twitter/X Feed Image
  'twitter:feed:image': {
    allowedMimeTypes: ['image/jpeg', 'image/png', 'image/gif', 'image/webp'],
    maxBytes: 5 * 1024 * 1024, // 5MB
    minWidth: 100,
    minHeight: 100,
    maxWidth: 4096,
    maxHeight: 4096,
    aspectRatioMin: 0.5,
    aspectRatioMax: 3.0,
  },

  // Twitter/X Feed Video
  'twitter:feed:video': {
    allowedMimeTypes: ['video/mp4'],
    maxBytes: 512 * 1024 * 1024, // 512MB
    minWidth: 32,
    minHeight: 32,
    maxWidth: 1920,
    maxHeight: 1200,
    aspectRatioMin: 0.5,
    aspectRatioMax: 2.0,
    durationMinSeconds: 0.5,
    durationMaxSeconds: 140, // 2 min 20 sec
  },

  // LinkedIn Feed Image
  'linkedin:feed:image': {
    allowedMimeTypes: ['image/jpeg', 'image/png', 'image/gif'],
    maxBytes: 8 * 1024 * 1024, // 8MB
    minWidth: 276,
    minHeight: 276,
    maxWidth: 4320,
    maxHeight: 4320,
    aspectRatioMin: 0.57,
    aspectRatioMax: 3.0,
  },

  // LinkedIn Feed Video
  'linkedin:feed:video': {
    allowedMimeTypes: ['video/mp4', 'video/quicktime', 'video/x-msvideo'],
    maxBytes: 200 * 1024 * 1024, // 200MB
    minWidth: 256,
    minHeight: 144,
    maxWidth: 4096,
    maxHeight: 2304,
    aspectRatioMin: 0.5625,
    aspectRatioMax: 2.4,
    durationMinSeconds: 3,
    durationMaxSeconds: 600, // 10 minutes
  },
}

/**
 * Gets the validation rule for a specific platform, placement, and media type.
 */
export function getClientValidationRule(
  platform: PlatformId | string,
  placement: Placement | string = 'Feed',
  mediaType: MediaType | string
): ClientMediaValidationRule | null {
  const key = `${platform.toLowerCase()}:${placement.toLowerCase()}:${mediaType.toLowerCase()}` as RuleKey
  return clientValidationRules[key] ?? null
}

/**
 * Pre-validates a file before upload using client-side rules.
 * Returns an array of error messages, or empty array if valid.
 */
export function preValidateFile(
  file: File,
  platform: PlatformId | string,
  placement: Placement | string = 'Feed'
): string[] {
  const errors: string[] = []

  // Determine media type from MIME
  const isImage = file.type.startsWith('image/')
  const isVideo = file.type.startsWith('video/')
  const mediaType: MediaType | null = isImage ? 'Image' : isVideo ? 'Video' : null

  if (!mediaType) {
    errors.push('Unsupported file type. Please upload an image or video.')
    return errors
  }

  const rule = getClientValidationRule(platform, placement, mediaType)
  if (!rule) {
    // No client rules defined - let backend handle it
    return []
  }

  // Check MIME type
  if (!rule.allowedMimeTypes.includes(file.type.toLowerCase())) {
    errors.push(`File type "${file.type}" is not supported for ${platform}. Allowed: ${rule.allowedMimeTypes.join(', ')}`)
  }

  // Check file size
  if (file.size > rule.maxBytes) {
    const maxMB = (rule.maxBytes / (1024 * 1024)).toFixed(1)
    const actualMB = (file.size / (1024 * 1024)).toFixed(1)
    errors.push(`File size (${actualMB}MB) exceeds maximum allowed (${maxMB}MB)`)
  }

  return errors
}

/**
 * Pre-validates image dimensions using client-side rules.
 * Call this after loading image metadata.
 */
export function preValidateImageDimensions(
  width: number,
  height: number,
  platform: PlatformId | string,
  placement: Placement | string = 'Feed'
): string[] {
  const errors: string[] = []
  const rule = getClientValidationRule(platform, placement, 'Image')

  if (!rule) return []

  // Check minimum dimensions
  if (width < rule.minWidth || height < rule.minHeight) {
    errors.push(`Image dimensions (${width}x${height}) are too small. Minimum: ${rule.minWidth}x${rule.minHeight}`)
  }

  // Check maximum dimensions
  if (width > rule.maxWidth || height > rule.maxHeight) {
    errors.push(`Image dimensions (${width}x${height}) are too large. Maximum: ${rule.maxWidth}x${rule.maxHeight}`)
  }

  // Check aspect ratio
  const aspectRatio = width / height
  if (aspectRatio < rule.aspectRatioMin || aspectRatio > rule.aspectRatioMax) {
    errors.push(
      `Aspect ratio (${aspectRatio.toFixed(2)}) is outside allowed range (${rule.aspectRatioMin.toFixed(2)} to ${rule.aspectRatioMax.toFixed(2)})`
    )
  }

  return errors
}

/**
 * Gets image dimensions from a File object using Image API.
 * Returns null if extraction fails.
 */
export function getImageDimensions(file: File): Promise<{ width: number; height: number } | null> {
  return new Promise((resolve) => {
    if (!file.type.startsWith('image/')) {
      resolve(null)
      return
    }

    const img = new Image()
    const url = URL.createObjectURL(file)

    img.onload = () => {
      URL.revokeObjectURL(url)
      resolve({ width: img.naturalWidth, height: img.naturalHeight })
    }

    img.onerror = () => {
      URL.revokeObjectURL(url)
      resolve(null)
    }

    img.src = url
  })
}

/** Error codes matching backend */
export const MediaValidationErrorCodes = {
  FileTooLarge: 'FILE_TOO_LARGE',
  UnsupportedMimeType: 'UNSUPPORTED_MIME_TYPE',
  DimensionsTooSmall: 'DIMENSIONS_TOO_SMALL',
  DimensionsTooLarge: 'DIMENSIONS_TOO_LARGE',
  AspectRatioInvalid: 'ASPECT_RATIO_INVALID',
  DurationTooShort: 'DURATION_TOO_SHORT',
  DurationTooLong: 'DURATION_TOO_LONG',
  FpsTooLow: 'FPS_TOO_LOW',
  FpsTooHigh: 'FPS_TOO_HIGH',
  UnsupportedContainer: 'UNSUPPORTED_CONTAINER',
  UnsupportedVideoCodec: 'UNSUPPORTED_VIDEO_CODEC',
  UnsupportedAudioCodec: 'UNSUPPORTED_AUDIO_CODEC',
  MetadataExtractionFailed: 'METADATA_EXTRACTION_FAILED',
  NoRulesForCombination: 'NO_RULES_FOR_COMBINATION',
} as const
